File: lib/engine/pnl.ts
// ============================================================================
// Aegis Vault — PnL Settlement
// Calculate unrealized PnL and settle (close) positions.
// ============================================================================

import { type Position, type TraderAccount, Side } from './types';

/**
 * Calculate unrealized PnL for a position at a given mark price.
 *
 * Long:  PnL = (markPrice - entryPrice) * size
 * Short: PnL = (entryPrice - markPrice) * size
 */
export function calculateUnrealizedPnl(
  position: Position,
  markPrice: number
): number {
  if (position.side === Side.Long) {
    return (markPrice - position.entryPrice) * position.size;
  } else {
    return (position.entryPrice - markPrice) * position.size;
  }
}

/**
 * Settle (close) a position at a given exit price.
 * Calculates realized PnL, adds/subtracts from collateral balance,
 * and removes the position from the account.
 *
 * Returns the realized PnL.
 */
export function settlePnl(
  position: Position,
  exitPrice: number,
  account: TraderAccount
): number {
  const realizedPnl = calculateUnrealizedPnl(position, exitPrice);

  // Add realized PnL to collateral (can be negative)
  account.collateralBalance += realizedPnl;

  // Return margin to collateral
  account.collateralBalance += position.marginAllocated;

  // Remove position from account
  const idx = account.positions.indexOf(position);
  if (idx !== -1) {
    account.positions.splice(idx, 1);
  }

  return realizedPnl;
}

/**
 * Calculate total unrealized PnL across all positions in an account.
 */
export function totalUnrealizedPnl(
  account: TraderAccount,
  markPrices: Map<string, number>
): number {
  return account.positions.reduce((total, pos) => {
    const markPrice = markPrices.get(pos.market) ?? pos.entryPrice;
    return total + calculateUnrealizedPnl(pos, markPrice);
  }, 0);
}

File: engine/src/pnl.rs
use std::collections::HashMap;
use crate::types::{Position, Side, TraderAccount};

pub fn calculate_unrealized_pnl(position: &Position, mark_price: f64) -> f64 {
    match position.side {
        Side::Long => (mark_price - position.entry_price) * position.size,
        Side::Short => (position.entry_price - mark_price) * position.size,
    }
}

pub fn settle_pnl(position: &Position, exit_price: f64, account: &mut TraderAccount) -> f64 {
    let realized_pnl = calculate_unrealized_pnl(position, exit_price);

    account.collateral_balance += realized_pnl;
    account.collateral_balance += position.margin_allocated;

    account.positions.retain(|p| p.market != position.market || p.side != position.side);

    realized_pnl
}

pub fn total_unrealized_pnl(account: &TraderAccount, mark_prices: &HashMap<String, f64>) -> f64 {
    account.positions.iter().map(|pos| {
        let mark_price = *mark_prices.get(&pos.market).unwrap_or(&pos.entry_price);
        calculate_unrealized_pnl(pos, mark_price)
    }).sum()
}
